plugins/cluster-etcd/src/main/java/org/opensearch/cluster/etcd/ETCDStateDeserializer.java

```
@@ -201,23 +209,105 @@ private static DataNodeState readDataNodeState(DiscoveryNode localNode, Client e
                 }
             }
             
-            for (int i = 0; i < futures.size(); i++) {
+            // Process the results
+            for (int i = 0; i < indexNames.size(); i++) {
                 String indexName = indexNames.get(i);
-                GetResponse getResponse;
-                try {
-                    getResponse = futures.get(i).get();
-                } catch (InterruptedException | ExecutionException e) {
-                    throw new RuntimeException(e);
+                IndexMetadata indexMetadata = buildIndexMetadataFromSeparateParts(
+                    indexName, 
+                    settingsFutures.get(i), 
+                    mappingsFutures.get(i)
+                );
+                indexMetadataMap.put(indexName, indexMetadata);
+            }
+        }
+        return new DataNodeState(localNode, indexMetadataMap, localShardAssignment);
+    }
+    
+    /**
+     * Builds IndexMetadata from separate settings and mappings etcd responses.
+     * Also populates constant values that don't need to be stored in etcd.
+     */
+    private static IndexMetadata buildIndexMetadataFromSeparateParts(
+        String indexName, 
+        CompletableFuture<GetResponse> settingsFuture, 
+        CompletableFuture<GetResponse> mappingsFuture
+    ) throws IOException {
+        Settings indexSettings = null;
+        MappingMetadata mappingMetadata = null;
+        
+        try {
+            // Fetch and parse settings
+            GetResponse settingsResponse = settingsFuture.get();
+            if (!settingsResponse.getKvs().isEmpty()) {
```

Here and on the mappings, I'm not sure if the empty check makes sense. 

If we fail to resolve one of the keys, I think we should throw an exception.